### ISSUE TITLE
API for nD shapes

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#296]
## 🚀 Feature
As discussed in #112 and other places, it will de desirable to extend support for our shapes layer to be `nD`.

One plan is to just support `nD` shapes were each shape is confined to a single (x, y) slice. For example the raw data maybe 4D, but a given shape will reside entirely within one timepoint and one z plane. If we restrict to supporting this case, how should be represent the data? Will we run into problems later when we imagine allowing rapid switching of which dimensions are the displayed ones (i.e. to see the `xz` projection?). Do we want to support a more `nD` like data model but somehow use projections? (it is not clear to me how to do this, but i'd be open to suggestions).

If we go with the above plan what should our API look like. Say you want to have a rectangle on z-slice 0 and a rectangle on z-slice 1. Do you pass
```python
[[[0, y0, x0], [0, y1, x1]],
 [[1, y2, x2],  [1, y3, x3]]]
```
where `[y0, x0]` and  `[y1, x1]` are the two corner points of the rectangle in the zeroth z-slice.
What happens though if someone tries to pass
```python
[[[0, y0, x0], [1, y1, x1]]]
```
Should we reject this rectangle as not supported? Or is now where we use some of our projection magic?

Internally do we want to have a mapping from z-slice -> 2D shape representation that allows us to use the `ShapeList` that we have now unchanged? I'd heavily favor this option is we're only supporting `nD` shapes in the `2D` way described above. If we want to support projections, then maybe it's more complicated.

Do @jni or @d-v-b have any thoughts?

**Comment by sofroniewn** [comment]
This is now implemented in #297 with an example at `examples/nD_shapes.py`

**Comment by sofroniewn** [comment]
This is now merged
